CIP-0109/README.md

```diff
@@ -0,0 +1,97 @@
+
+## Specification
+Modular exponentiation is mathematically defined as the equivalence relation
+
+$$
+a \equiv b^{e} (\bmod{m})
+$$
+
+Here we have that the base $b$ is an integer, the exponent $e$ a non-negative integer and the modulus $m$ a positive integer. To be more specific, we want $e$ to be larger or equal to zero, and $m$ is larger than zero. This is because what we are effectively doing, is multiplying $e$ copies of $b$ and a modulo reduction by $m$. In this context, multiplying a negative number of copies of $b$ has no definition.
+
+
+$$
+\begin{equation}
+a \equiv \underbrace{(b \times b \times ... \times b)}_{e \textrm{ times}} (\bmod{m} )
+\end{equation}
+$$
+
+That said, in the context of multiplicative inversion, a negative exponent can be interpreted as taking the exponent of the inverse of the base number. That is
+
+$$
+\begin{equation}
+b^{-e} (\bmod{m}) := (b^{-1})^{e} (\bmod{m})
+\end{equation}
+$$
+
+We propose to also include this extension to the plutus built-in, for optimized inversion when this is possible. This is inversion is not guaranteed to exist for all numbers $b$, only when the modulus $m$ is a prime number this is guaranteed. We also propose that the built-in fails if this inverse does not exist, and if a modulus is provided that is smaller than one.
+
+### Function definition
+With the above, we define a new Plutus built-in function with the following type signature
+
+```hs
+modularExponentiation :: Integer -> Integer -> Integer -> Integer
+```
+here the first argument is the base, the second the exponent and the third the modulus. As mentioned above, the behavior of this function is that it fails if the modulus is not a positive integer, or if the inverse of the base does not exist for a negative exponent. For the lower level implementation, we propose the usage of the `integerPowMod` function in the `ghc-bignum` packages. This function has the desired functionality, is optimized, and is easy to integrate in the plutus stack.
+
+### Cost model
+The computational impact of modular exponentiation is complexified by it having three arguments. That said, observe that the integers used can always be bound by the modulus. Preliminary [benchmarks](https://github.com/perturbing/expFast-bench/tree/cca69b842050de9523493d52c20384bc50c80b22) on the time consumption of this `integerPowMod` function show that it can be costed constant in the size of its first argument (the base) and linear in the other two.
+
+## Rationale: how does this CIP achieve its goals?
```

Comment:
Would be nice to see _a little more_ discussion in the `Rationale`.

Perhaps:
- What would be other alternatives to this design?
- Any backwards compatibility considerations?

Reply:
Mainly resolved with https://github.com/cardano-foundation/CIPs/pull/641/commits/ebcd2eced90c7b98f26eebf8e776808201c94335.